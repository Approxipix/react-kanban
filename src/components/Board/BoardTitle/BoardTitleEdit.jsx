import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { editBoardTitle } from '../../../redux/boardReducer/actions';
import ClickOutside from '../../ClickOutside';
import { Input, } from '../../BaseComponent';
import styled from 'styled-components'

const Form = styled.form`
  max-width: calc(100vw - 12rem);
`;

class BoardTitleEdit extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: props.boardTitle || '',
    };
  }

  handleChange = (value) => {
    this.setState({
      title: value,
    });
  };

  handleSubmit = (e) => {
    e.preventDefault();
    const { title } = this.state;
    if (!title)  return;
    this.props.actions.editBoardTitle({
      boardId: this.props.boardId,
      boardTitle: title
    });
    this.props.toggleEditTitle()
  };

  render() {
    const { title } = this.state;
    let inputWidth = (title.length + 1) * 13;
    return (
      <ClickOutside toggleOpened={this.props.toggleEditTitle}>
        <Form onSubmit={this.handleSubmit}>
          <Input
            type="text"
            size={'1.3rem'}
            weight={'600'}
            width={inputWidth + 'px'}
            value={title}
            placeholder="Edit board title"
            onChange={(e) => this.handleChange(e.target.value)}
            onBlur={this.handleSubmit}
            onKeyPress={(e) => inputWidth = ((e.target.value.length + 1) * 13)}
            autoFocus
            spellCheck={false}
          />
        </Form>
      </ClickOutside>
    );
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      editBoardTitle: editBoardTitle,
    }, dispatch)
  };
}


export default connect(null, mapDispatchToProps)(BoardTitleEdit);
