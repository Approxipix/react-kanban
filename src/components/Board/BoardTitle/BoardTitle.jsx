import React, { Component } from 'react';
import PropTypes from "prop-types";
import { connect } from "react-redux";
import BoardTitleEdit from "./BoardTitleEdit";
import styled from "styled-components";

const Title = styled.h1`
  max-width: calc(100vw - 12rem);
  margin: 0;
  padding: .5rem;
  font-size: 1.3rem;
  color: #fff;
  font-weight: 600;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
`;
Title.displayName = 'Title';

class BoardTitle extends Component {
  constructor(props) {
    super(props);
    this.state = {
      editTitle: false,
    }
  }

  toggleEditTitle = () => {
    this.setState({
      editTitle: !this.state.editTitle,
    })
  };

  render() {
    const { boards, currentBoardID } = this.props;
    const { editTitle } = this.state;

    const board = boards[currentBoardID];
    if (!board) return;
    return (
      !editTitle ? (
        <Title onClick={() => this.toggleEditTitle()}>
          {board.title}
        </Title>
      ) : (
        <BoardTitleEdit
          boardId={board._boardId}
          boardTitle={board.title}
          toggleEditTitle={this.toggleEditTitle}
        />
      )
    )
  }
}

BoardTitle.propTypes = {
  currentBoardID: PropTypes.string.isRequired,
  boards: PropTypes.objectOf(
    PropTypes.shape({
      _boardId: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      color: PropTypes.string,
      lists: PropTypes.arrayOf(PropTypes.string),
    })
  ).isRequired,
};

function mapStateToProps(state) {
  return {
    boards: state.boardReducer,
    currentBoardID: state.rootReducer.currentBoardID,
  }
}

export default connect(mapStateToProps)(BoardTitle);
