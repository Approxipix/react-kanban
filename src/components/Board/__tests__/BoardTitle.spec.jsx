import React from 'react';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import { BoardTitle } from '../BoardTitle/BoardTitle';

describe('<BoardTitle>', () => {
  let state;
  let component;

  beforeEach(() => {
    state = {
      editTitle: false,
    };
    const props = {
      board: {
        title: 'Board Title'
      },
    };

    component = shallow(<BoardTitle {...props} />);
    component.setState({...state});
  });

  it('should render title if editTitle is false', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render BoardTitleEdit component if editTitle is true', () => {
    component.instance().toggleEditTitle();
    expect(toJson(component)).toMatchSnapshot();
  });

  it('checks if toggleEditTitle sets editTitle correctly', () => {
    component.instance().toggleEditTitle();
    expect(component.instance().state.editTitle).toEqual(true);

    component.instance().toggleEditTitle();
    expect(component.instance().state.editTitle).toEqual(false);
  });

  it("should call toggleEditTitle after click on title", () => {
    const spy = jest.spyOn(component.instance(), "toggleEditTitle");

    component.find('Title').simulate('click');

    expect(spy).toHaveBeenCalledTimes(1);
  });
});
