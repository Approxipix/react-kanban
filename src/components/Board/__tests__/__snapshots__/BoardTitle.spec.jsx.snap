// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BoardTitle> should render BoardTitleEdit component if editTitle is true 1`] = `
<Connect(BoardTitleEdit)
  toggleEditTitle={[Function]}
/>
`;

exports[`<BoardTitle> should render title if editTitle is false 1`] = `
<Title
  onClick={[Function]}
/>
`;
